src/response/error.rs

```diff
@@ -4,108 +4,124 @@ use tonic::codegen::http;
 
 use crate::auth::AuthError;
 
+#[derive(Debug)]
+pub enum MomentoErrorCode {
+    /// Invalid argument passed to Momento client
+    InvalidArgumentError,
+    /// Service returned an unknown response
+    UnknownServiceError,
+    /// Resource with specified name already exists
+    AlreadyExistsError,
+    /// Cache with specified name doesn't exist
+    NotFoundError,
+    /// An unexpected error occurred while trying to fulfill the request
+    InternalServerError,
+    /// Insufficient permissions to perform operation
+    PermissionError,
+    /// Invalid authentication credentials to connect to service
+    AuthenticationError,
+    /// Request was cancelled by the server
+    CancelledError,
+    /// Request rate, bandwidth, or object size exceeded the limits for the account
+    LimitExceededError,
+    /// Request was invalid
+    BadRequestError,
+    /// Client's configured timeout was exceeded
+    TimeoutError,
+    /// Server was unable to handle the request
+    ServerUnavailable,
+    /// A client resource (most likely memory) was exhausted
+    ClientResourceExhausted,
+    /// System is not in a state required for the operation's execution
+    FailedPreconditionError,
+    /// Unknown error has occurred
+    UnknownError,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("{details}")]
+pub struct MomentoGrpcErrorDetails {
+    pub code: tonic::Code,
+    pub details: Cow<'static, str>,
+    pub metadata: Option<tonic::metadata::MetadataMap>,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("{message}")]
+pub struct SdkError {
+    pub message: Cow<'static, str>,
+    pub error_code: MomentoErrorCode,
+    #[source]
+    pub inner_exception: Option<ErrorSource>,
+    pub details: Option<MomentoGrpcErrorDetails>,
+}
```

Comment:
Instead of specifying the fields for each MomentoError enum case, I made a struct that each error should adhere to instead (with the exception of Miss and TypeError, weren't sure about those).

Not exactly certain about the types specified here but these fields are at least consistent with the other SDKs

Reply:
I like this but then I think the way you're defining the stuff in the enum below, you are basically defining each Error type as a 1-item-tuple. (I think. I'm still learning some of this rust syntax myself.)

For such tuples I've seen some weird syntax where you'd have to write something like `error.0.description` to access the description field, and we don't want that. But in your case since it's an enum they might have to do a `match` first, and maybe they would be destructuring the inner error in the `match`, so maybe it's okay. is there some code somewhere in this PR that illustrates what it looks like to handle one of these errors and access one of the fields inside?

I'm also wondering about whether it's possible for someone to have a reference to a MomentoError and then access the inner fields without having to `match` to a specific value from the enum first. I wonder if we can do like `impl MomentoError { ... }` and provide some common accessors that can be used without the match. I don't think it works that way but it'd be really desirable from an end user perspective so we might want to play around and look for some way to pull that off. Like, maybe `MomentoError` becomes a struct with the fields that you currently put into `SdkError`, and then we add a field to it called `error_type` or something, and the value of that is the enum?

We can chat about this and trade ideas if you like.

cc: @kvcache you will probably have opinions about this part.

Reply:
Been testing locally to get an idea of the current UX. Was able to use [`if let`](https://doc.rust-lang.org/book/ch06-03-if-let.html) to match on the specific MomentoError enum type and then get at the SdkError struct by creating another variable:

```
let result = cache_client.create_cache(CACHE_NAME.to_string()).await;
match result {
    Ok(_) => println!("Successfully created cache '{}'", CACHE_NAME),
    Err(e) => if let MomentoError::AlreadyExists(already_exists) = e {
        println!("Cache already exists: {} | Metadata: {:?}", already_exists.message, already_exists.details.unwrap().metadata);
    } else if let MomentoError::NotFound(not_found) = e {
        println!("Cache '{}' not found | Metadata: {:?}", not_found.message, not_found.details.unwrap().metadata);
    } else {
        println!("Error: {}", e);
    }
}
```